### ISSUE TITLE
Implement SIMD where applicable

### ISSUE THREAD

**Issue by olback** [RustCrypto/RSA#49]
What got me looking into this was the somewhat slow key generation and the related issue #29. A solution or rather, improvement would be to implement SIMD, [*Single Instruction Multiple Data*](https://en.wikipedia.org/wiki/SIMD). I found a paper by freescale semiconductor on this topic here: http://application-notes.digchip.com/314/314-66328.pdf.

Since all processors do not support the AVX/SSE/SIMD family of instructions, this would have to be implemented under a feature flag or as in the case of [aes_gcm](https://docs.rs/aes-gcm/0.5.0/aes_gcm/), the feature is enabled when the compiler is passed these flags:
```
RUSTFLAGS="-Ctarget-cpu=sandybridge -Ctarget-feature=+aes,+sse2,+sse4.1,+ssse3"
```

I might take a deeper look at this when my exams are over.

**Comment by zicklag** [comment]
I would suggest looking/asking around about the best ways to support SIMD in Rust, too, because I think there are crates that allow for determining whether or not to use SIMD at runtime based on CPU features and such.

**Comment by tarcieri** [comment]
Yes, there's a stable [std::is_x86_feature_detected](https://doc.rust-lang.org/std/macro.is_x86_feature_detected.html) macro as well as the [`cpuid-bool`](https://docs.rs/cpuid-bool/0.1.2/cpuid_bool/) crate. There's also a nightly-only [std::is_arm_feature_detected](https://doc.rust-lang.org/std/macro.is_arm_feature_detected.html) macro. But these are only helpful after you have a SIMD backend in the first place.

Another crate to look at is [`packed_simd`](https://github.com/rust-lang/packed_simd), which is an implementation of the [`std::simd` proposal](https://github.com/gnzlbg/rfcs/blob/ppv/text/0000-ppv.md). Unfortunately, it's also nightly-only.

**Comment by burdges** [comment]
Any idea what is optimal here?  We'd first want some fixed size bigint crate that avoids the `Vec` used by num-bigint?  We'd then do SIMD support only for that crate?  Or something more bespoke?

**Comment by tarcieri** [comment]
> We'd first want some fixed size bigint crate that avoids the Vec used by num-bigint? 

This is what we wrote `crypto-bigint` for: https://github.com/rustcrypto/crypto-bigint

Should have another release soon, although it doesn't yet have any SIMD features.

One thing I can try to do prior to the next release is add support for loading certain sizes of big ints (it's using const generics internally, but still allows specialization around size) into SIMD registers, which will at least leave the door open for SIMD optimizations.
